### ISSUE TITLE
Steady state phase of target in SyncAction

### ISSUE THREAD

**Issue by johannesWen** [esmini/esmini#51]
I found another useful feature for the SyncAction. It would be handy if you could define a steady state phase (constant speed) of the target during the SyncAction, either by adding another position (start of the steady state) or by a relative time to the synchronization point. 
The background is that EURO NCAP scenarios very often require this steady state phase. In most cases a time to collision of four seconds is specified as the start condition of a test and during these four seconds the speed of the vehicles must remain constant. 
This might be a useful extension of OpenScenario version 1.1. 
What do you think?

**Comment by eknabevcc** [comment]
I think this is a very interesting and good idea. I have received a similar idea before, but in that case the proposal was to launch a pedestrian at exact right time in order to reach the destination with specified constant speed. I guess it's very similar to, maybe even a special case of, your first variant - just that the SyncronizeAction starts from that position, not using it as a intermediate position. 

Actually, your proposed feature - including the two options (position or remaining time) would be a great add-on I believe! 

But we need to think about how to express it in terms of the XSD in a reasonable way. If we look at current proposal, including the tolerances, do you have any ideas/preferences how to add the SteadyState concept to it, also maintaining backward compatibility?

```
<xsd:complexType name="SynchronizeAction">
    <xsd:all>
        <xsd:element name="TargetPositionMaster" type="Position"/>
        <xsd:element name="TargetPosition" type="Position"/>
        <xsd:element name="FinalSpeed" type="FinalSpeed" minOccurs="0"/>
    </xsd:all>
    <xsd:attribute name="masterEntityRef" type="String" use="required"/>
    <xsd:attribute name="targetToleranceMaster" type="Double" use="optional"/> 
    <xsd:attribute name="targetTolerance" type="Double" use="optional"/> 
</xsd:complexType>
```

I guess that the SteadyState features should be optional to start with, and that you could choose either to specify a point (element of type="Position") or a distance (attribute of type="double"). I'm not sure how to express that in XSD.

**Comment by johannesWen** [comment]
I thought a bit about a possible xml-schema. How about this version. I'm not quite sure about backward compatibility.
```
<xsd:complexType name="SynchronizeAction">
	<xsd:all>
		<xsd:element name="TargetPositionMaster" type="Position"/>
		<xsd:element name="TargetPosition" type="Position"/>
		<xsd:element name="FinalSpeed" type="FinalSpeed" minOccurs="0"/>
		<xsd:element name="SteadyState" minOccurs="0">
			<xsd:complexType>
				<xsd:choice>
					<xsd:element name="targetPositionSteadyState" type="Position"/>
					<xsd:element name="targetDistanceSteadyState" type="Double"/>
					<xsd:element name="targetTimeSteadyState" type="Double"/>
				</xsd:choice>
			</xsd:complexType>
		</xsd:element>
	</xsd:all>
	<xsd:attribute name="masterEntityRef" type="String" use="required"/>
	<xsd:attribute name="targetToleranceMaster" type="Double" use="optional"/>
	<xsd:attribute name="targetTolerance" type="Double" use="optional"/>
</xsd:complexType>
```

The idea is to have an optional element 'SteadyState' and in this element you can choose between an initial position of the steady state phase, a remaining distance to the SyncPoint or a remaining time to the SyncPoint.

**Comment by eknabevcc** [comment]
I've implemented and committed [(8947671)](https://github.com/esmini/esmini/commit/8947671d2462acc78a5d8f08f82af668d700145a) a draft version of this interesting feature. I see it as an experimental extension to OSC 1.0 that we can try out before finalizing proposal to OSC 1.1. 
Based on your proposed XSD I did a few minor adjustments, e.g. moving the SteadyState stuff in under FinalSpeed, as I guess it only makes sense in combo with that element. 
Please check [docs/osc-extensions.xml](https://github.com/esmini/esmini/blob/master/docs/osc-extensions.xml) for updated XSD proposal. (I'm not sure if it is correct XSD, but hopefully we're getting close).
And I updated [resources/xosc/pedestrian_traj_synch.xosc](https://github.com/esmini/esmini/blob/master/resources/xosc/pedestrian_traj_synch.xosc) to include TargetDistanceSteadyState as an example.

**Comment by johannesWen** [comment]
Looks good to me. 😃 Yes you are right, the final speed is necessary for this steady state feature.
If you could make a new build with this experimental extension, I will test it with my scenario and give you a feedback.

**Comment by eknabevcc** [comment]
There's not a new release yet, but hopefully you can grab updated executables from latest build here:
[https://ci.appveyor.com/project/esmini/esmini/builds/36574227](https://ci.appveyor.com/project/esmini/esmini/builds/36574227)

1. Click on job name for relevant platform
2. Then you find the binary file package under tab "Artifacts"

Note that theses packages only includes binary files, no resources like scenarios or models.

BTW: The pos option is a bit tricky. How it is implemented right now is like a variant of distance: esmini will measure the distance from the steady state position to the destination position - and use it just as the distance option. Regardless of the position laying on the path to destination or elsewhere. I think that is the most straight forward solution - however needs to be explained for the user... 
Other interpretation would be to add it as a destination - but then it gets complicated. What if it's not along path? And that's not always easy to check on beforehand, e.g. when driver in the loop.
What do you think? (I thought about removing it as an option- but I still think it can be really useful as is)

**Comment by johannesWen** [comment]
Alright, I will test it soon 👍 

I would say that the creator of the scenario has to care about the correct steady state position, otherwise the scenario would not make sense if the position is never reached.
Do you use the "targetTolerance" value here to check if the steady state position is reached?

**Comment by eknabevcc** [comment]
Actually I just check the distance. A way of putting it: The steady state position will define an outer circle around destination - as soon as object enters that area the steady state is applied. Then the actual steady state position does not need to lie on the path...
![image](https://user-images.githubusercontent.com/29146839/100600027-095ce980-3301-11eb-882c-60211cf42f85.png)
What do you think? Does it makes sense or would it be better with an actual position that the object need to touch (within tolerances) on its way to destination?
The time and distance attribute options are much simpler and more easy to explain. So a lazy solution would be to remove the Position option :)

**Comment by johannesWen** [comment]
Yes that looks reasonable. I think this is a good and more robust approach. 👍 

In my use case of the CPTA scenario it's a bit more challenging anyway. Here the Ego vehicle is moving along a trajectory and I have to consider the remaining distance along the trajectory. I was able to create the scenario successfully by calculating the synchronization points of the ego and target accordingly in advance. I get quite good results, as you can see on the following slide.
![grafik](https://user-images.githubusercontent.com/57130737/100609088-b6d5fa00-330d-11eb-9053-6355a05c3454.png)

Nevertheless, I would find a steady state synchronization position extremely helpful. 😃

**Comment by johannesWen** [comment]
Hi Emil,
I tested your steady state - implementation of the "TargetDistanceSteadyState" and "TargetTimeSteadyState". It looks good to me.
![grafik](https://user-images.githubusercontent.com/57130737/100997736-f2b1cf00-355a-11eb-8561-e8d6c35ce43f.png)
![grafik](https://user-images.githubusercontent.com/57130737/100997770-fe04fa80-355a-11eb-94cf-21827a6a41f6.png)

Concerning, the "TargetPositionSteadyState" I have a question. Is it a position of the master or the target? If I use a steady state position on the trajectory of the target, I get a strange behavior..
![grafik](https://user-images.githubusercontent.com/57130737/101018481-da967b80-356b-11eb-8703-6272c2c052ae.png)

**Comment by eknabevcc** [comment]
Hi Johannes. Sorry for delay. Thanks for testing the SteadyState and good to hear that Time and Distance seems OK.
Regarding Position. It's a bit tricky. First, it's related to target (the vehicle being controlled = the object to which the action is assigned). And the idea was to just convert it into a Distance case - using the distance from the Position to the target position (destination) of the Target vehicle. But maybe there is a bug in the implementation on my side. 

But actually, while refining the proposal for 1.1 I had some struggle explaining, in a clear way, the Position case. So I'm thinking about skipping that option for now. The time to 1.1 deadline is short, and I see a risk of introduce something not optimal - which then will be tricky to change (keeping backward compatibility, which always is a goal). By waiting, we can fine tune the concept and introduce the Position as a third option in 1.2. That's my current thinking. What do you think about that?

I would like to have your opinion: Do you find the Position case crucial/important, or do you think that Distance and Time is useful enough, at least as a first step?

**Comment by johannesWen** [comment]
Hi, no problem. Yes, I agree with you. For now, the distance- and time steady state definitions are quite sufficient. I still think the position definition would be useful, but it is not that urgent. At the moment I am very happy with the time and distance extensions.

It would also be helpful, if the measurement of the distance would be done along the trajectory. If one is used. I am not sure. If a road network is used, is the distance calculated along the road?

**Comment by eknabevcc** [comment]
Sounds good 👍 Let's start with Distance and Time. It has been proposed, we'll see if it makes it into the standard and to what version (1.1 or later). 
I see esmini as a good sandbox for testing new ideas and features. And I like this feature, so we'll keep it with all three options including Position in esmini regardless.

Regarding distance measurement; The way it SHOULD work is as follows:
`If a valid route is found from the scenario object to its corresponding target position, either through road network or along a trajectory, then the distance is calculated along the identified route. Otherwise the Euclidean distance is applied.
`
However, currently in esmini, it works just like you guessed:
When a valid route can be resolved to the target position through the road network, the distance is calculated along the route. Otherwise, if a route can't be resolved or on a trajectory, the Euclidean distance is used.

But definitely the plan is to calculate distance along trajectory as well. But there is an issue we need to fix first: Currently there can be only one shape in a trajectory. Which means that users need to put multiple actions and actually events in order to concatenate multiple shapes, e.g. clothoids, into a combined trajectory. Making it really hard to calculate distance. So we have ideas on how to solve this limitation first, then we'll be able to calculate distance along any trajectory (single or concatenated). No time plan though, but I'll keep you posted.

**Comment by johannesWen** [comment]
Hi, I figured out a problem with the newest build (e4648b440c2e9c14d3005bbdfb755b8f11952b18). When I run the 'synch_with_steady_state.xosc' scenario, the simulation freezes after some time. The log file shows no warning or error. I am sure you have an idea what the problem is ;-)

**Comment by eknabevcc** [comment]
Thanks for info. Probably a bug that I want to find and eliminate. I really want to reproduce it, but failed so far (test on Win and Linux). Questions:
1. Did you run the executable from one of the release zip-packages? Or did you build yourself, in that case what OS and compiler? Do you use provided libs (those downloaded automatically by cmake) for dependencies?
2. Do you have any special setup, e.g. adding any controller (or other modification of .xosc file) or applying some flags, e.g. --trails, --camera_mode ...

**Comment by johannesWen** [comment]
Hm, I just tested it on another Windows machine but there the bug does not occur..
 I used the "run_synch_with_steady_state.bat" from the release zip. It is probably a problem of my other machine.

**Comment by eknabevcc** [comment]
Hm, interesting. Would be nice to track down that bug.
Does it always happen at same time in the scenario? In that case, at about what time (maybe related to some step in SynchronizeAction...)? Don't need an exact millisecond, second precision is great enough.

**Comment by johannesWen** [comment]
It always happens at 0.2 s - 0.4 s in the "synch_with_steady_state.xosc" scenario. The bug also happens in my crossing scenarios and the "highway_merge_advanced.xosc" scenario. It is very strange, because it often occurs at a different time in the highway merge scenario.
It looks like that the scenarios run without errors, because the log files have a normal output. Only the visualization freezes.

**Comment by eknabevcc** [comment]
Good that at least you can re-create the bug. And good that it's not only SynchronizeAction. 
I would very much appreciate if you could support identifying the bug. My plan would be to, in a few steps, remove features that potentially could cause the freeze. And you could just test and see if issue still present or not.
Would that be possible?
In that case we just should sort out a few questions:
Can I commit this special version on a branch and you pull from that and build yourself?
Or do you want me to build? That's totally OK, just need to know whether to build for Windows or Linux?

**Comment by eknabevcc** [comment]
Wait a minute: Exact those two run-scripts selects the flex-orbit camera. So hopefully I can narrow the bug-searching to that feature. 

Could you please try the following: Edit the two bat files replace `--camera_mode flex-orbit` with `--camera_mode orbit`.
Would be very interesting to hear the result, does it still freeze or not?

**Comment by johannesWen** [comment]
Yes, I can try to identify the bug and remove features to see which one produces the freeze bug. If you could build this branch for windows, that would be great.

I also tried this, but it always freezes, no matter which camera_mode I use..

**Comment by eknabevcc** [comment]
Ok. Just to clarify, I don’t expect you to hunt the bug in code. I have a few ideas that I can commit to a branch, and then you can just test and report result.
But, of course, if you have some ideas - don’t hesitate to tell me or implement on your own. We could use same or different branches, we’ll sort out that part.

**Comment by eknabevcc** [comment]
I created a branch: https://github.com/esmini/esmini/tree/freeze_issue for our bug hunt
It's based on bb22d5a, which below is referred to as original.

[freeze_issue.zip](https://github.com/esmini/esmini/files/5742868/freeze_issue.zip) contains two esmini executables:
1. original
2. modified_v1: 
    - Disable trail geometries 
    - Disable dynamic transparency mode for VisibilityAction
    - Enforce OSG single thread mode

Please test both. Easiest way is to just replace the esmini.exe in your bin folder. 

The original should be very similar to what AppVeyor built for last release. So that should probably fail. But IF original works, then we should focus on differences between my and AppVeyor builds. 

My expectations is that original should definitely fail, and maybe 70% chance of v1 to fail as well. But it's a first step hunting the bug.

**Comment by johannesWen** [comment]
I tested both versions of the .exe but unfortunately the "synch_with_steady_state.xosc" scenario still freezes. 
What I found out is, that when I use the random vehicles as fall back if no catalog vehicle is found, the scenario do not freeze. I am not sure if this helps.

**Comment by eknabevcc** [comment]
I don't know if it's good or bad that it still freeze. At least the good news is that the bug is not likely in those disabled parts... 
The info regarding random vehicles, I assume you mean the colored boxes, is very interesting. 
There is separate wheel nodes in the car_white.osgb model, which then is animated/rotated (roll and heading) during simulation. One small test step would be to remove that very model.
I just committed a slight change to the VehicleCatalog, so that car_white is represented by car_yellow.osgb model. Just to take the wheel nodes away. Can you please pull and test again with the catalog in place?

**Comment by johannesWen** [comment]
Unfortunately it still freezes with the yellow car…
This is very strange because the cut-in scenario runs without a problem and there is the white/yellow car.

**Comment by eknabevcc** [comment]
Yes. Strange. I'm thinking about what can be difference between generated "box car" and loaded .osgb model. A silly test would be to remove (or rename) the shadow_face.osgb file to prevent it from being loaded.
That's like a single face, with texture including alpha channel, put under the objects to mimic shadow. It's such a long shot but on the other hand an easy test, to see if that one could influence the execution... Please try.

**Comment by johannesWen** [comment]
Hm, no. Does not help.
What helps, at least to some point, is if I delete the .osgb of the road network. Then the simulation runs till after the junction and then freezes.
![grafik](https://user-images.githubusercontent.com/57130737/104556512-05cabb00-5640-11eb-8f50-a73de0267a6a.png)

**Comment by eknabevcc** [comment]
Sorry for long pause. I hope you still want to support finding the strange issue.
In that case: I've just pushed [a version](https://github.com/esmini/esmini/commit/357863e4f7b086ccd1395ab9fc79b497ab464eb7) on our branch with further things removed/disabled:
- Don't create OpenDRIVE OSG road features (lines and dots)
- Don't create road sensor objects (probably not activated anyway in this scenario, but now totally removed from code)
- Don't create and show the on-screen info text
- Don't enforce OSG SingleThread mode

The last bullet means that OSG will choose threading model. Probably it will not choose SingleThread, but one of multiple variants of multi-threading. You can always find out the mode by pressing 's' key twice (see pic below). And You can force it into SingleThread by providing the command line argument `--SingleThreaded`. Please test with and without that argument. (Also, when not specifying --SingleThreaded, you can switch between various modes with 'm' key - you can try that also. If it seems to have any effect I can build separate versions locking into a specific mode, could be a next step).

![image](https://user-images.githubusercontent.com/29146839/106566196-47a89c00-6530-11eb-9c89-0e6095ef4103.png)

**Comment by johannesWen** [comment]
Hey, no problem. I am still interested and I have good news. I guess I found the problem.
First I tested your new version and the scenario worked. Then I tested the scenario with the 2.1.5 build and I deactivated some features.
So when I deactivate the on-screen info (--info_text off), the scenario does not freeze. If I activate the on-screen text with 'i', it freezes again.

**Comment by eknabevcc** [comment]
:) Finally pay-back, something to focus on! Thanks for persistent testing. Let's hope 1. info text is the actual root cause and 2. that it can be fixed. 
Actually I've seen some flickering of that text, which has been a bit irritating and maybe related. So I look forward to dig in and tackle it asap, hopefully this or next week.

**Comment by eknabevcc** [comment]
I tried a few low hanging things, like updating the info text in a dedicated OSG-callback. And also changing some OSG flags. But no visible improvements - still flickering, and probably issue persists. :(
Just wanted to let you know I didn't forget about this issue. I will (like a background process) continue looking for a solution that also makes the info text stable.

**Comment by johannesWen** [comment]
I'm glad to hear that the proposal made it into the new version. 😃
